src/components/feedback/Dialog.tsx

```diff
@@ -48,19 +49,33 @@ type ComponentProps = {
    * Optional TransitionComponent for open (mount) and close (unmount) transitions
    */
   transitionComponent?: TransitionComponent;
-
-  /**
-   * Dialog lays out its content in a Panel by default, but `custom` allows any
-   * layout of Dialog children
-   */
-  variant?: 'panel' | 'custom';
 };
 
 // This component forwards a number of props on to `Panel` but always sets the
 // `fullWidthHeader` prop to `true`.
-export type DialogProps = PresentationalProps &
+export type PanelDialogProps = PresentationalProps &
   ComponentProps &
-  Omit<PanelProps, 'fullWidthHeader'>;
+  Omit<PanelProps, 'fullWidthHeader'> & {
+    /** Dialog lays out its content in a Panel by default */
+    variant?: 'panel';
+  };
+
+type NonPanelDialogProps = PresentationalProps &
+  ComponentProps & {
+    /** `custom` allows any layout of Dialog children */
+    variant: 'custom';
+
+    /** @deprecated Title is ignored in custom dialogs */
+    title?: string;
+  };
+
+export type DialogProps = PanelDialogProps | NonPanelDialogProps;
+
+function isPanelProps(
+  props: DialogProps,
+): props is Omit<PanelProps, 'fullWidthHeader'> {
+  return props.variant !== 'custom' && 'title' in props;
```

Comment:
I think that checking whether this is a panel should be based exclusively on the `variant` prop. Otherwise if, for any reason, the title attribute is incorrectly missing at runtime (eg. due to something that slips by TypeScript), the code will confusingly go down the wrong path.